on-chain/fee-distribution/contracts/FeeDistributor.sol

```diff
@@ -0,0 +1,217 @@
+
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+
+contract FeeDistributor is Ownable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using SafeMath for uint256;
+
+    address public want;
+    address public strategy;
+
+    struct Distribution {
+        uint256 totalFees;
+        mapping (address => uint256) fractions;
+        mapping (address => uint256) fundsWithdrawn;
+        address[] users;
+    }
+
+    mapping (uint256 => Distribution) private distributions;
+    mapping (address => uint256) private currentWithdrawalDistributionId;
+    uint256 public distributionIndex;
+
+    /**
+     * @dev The constructor sets initial values
+     * @param _want address The address of coin
+     * @param _strategy address The address of linked strategy
+     * @param userAddresses array of address of users that can withdraw fees
+     * @param fractions array of uint256 that will contain fractions of fees distribution
+     */
+    constructor (
+        address _want,
+        address _strategy,
+        address[] memory userAddresses,
+        uint256[] memory fractions
+    ) public {
+        require (_want != address(0));
+        require (_strategy != address(0));
+
+        want = _want;
+        strategy = _strategy;
+
+        uint256 fractionsLength = fractions.length;
+        require (userAddresses.length == fractionsLength, "Error: Length mismatch");
+
+        distributionIndex = 0;
+        distributions[distributionIndex].users = userAddresses;
+        for (uint256 i = 0; i < fractionsLength; i++) {
+            distributions[distributionIndex].fractions[userAddresses[i]] = fractions[i];
+        }
+    }
+
+
+    /**
+     * @dev Updates fractions for each user
+     * @param userAddresses array of address of users that can withdraw fees
+     * @param fractions array of uint256 that will contain fractions of fees distribution
+     * @return true if updation successful
+     */
+    function updateFractions(
+        address[] calldata userAddresses,
+        uint256[] calldata fractions
+    ) external onlyOwner returns (bool) {
+        uint256 fractionsLength = fractions.length;
+        require (userAddresses.length == fractionsLength);
+
+        distributionIndex = distributionIndex.add(1);
+        distributions[distributionIndex].users = userAddresses;
+        for (uint256 i = 0; i < fractionsLength; i++) {
+            distributions[distributionIndex].fractions[userAddresses[i]] = fractions[i];
+        }
+    }
+
+    /**
+     * @dev called externally by strategy to transfer fees to this contract
+     * @param amount uint256 the amount that has to be sent to this contract( for distribution)
+     * @return true if successful
+     */
+    function sendFees(uint256 amount) external returns (bool) {
```

Comment:
If we didn't have multiple periods then totalFee accumulated before any withdrawals would just be token.balanceOf(this).
With some withdrawals, totalFee would be token.balanceOf(this)  + totalWithdrawals
With multiple periods totalFee in current period is balanceOf(this) + totalWithdrawals(this period) - sum of previous period total fees.

Reply:
So what we could do is: add state vaiable for currentTotalFees, currentTotalWithdrawals, sumPreviousTotalFees and update each of these whenever we need to create a new period (which we still store in array).  
i.e. basically same approach as current for previous period but break out the current period into separate state variables so that we can just use the token balanceOf function to determine fees accrued this period  *so that* we can just transfer tokens to this contract for fee distribution *so that* the strategy contract is then agnistic to whether it is sending to contract or EOA